rework monaco commands to align with VS Code

akosyakov opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

- Reverts https://github.com/eclipse-theia/theia/commit/40b1d45d45fdfc6b1db47c0690e450c5b28eb467 to align with VS Code behaviour as before
- Refactors monaco-commands to fix Find/Replace/Undo/Redo/Select All, but at the same time without breaking previous behaviour: https://github.com/eclipse-theia/theia/blob/42ea5bd196d6efa579ed571e64f7fc2e5e269e19/packages/monaco/src/browser/monaco-command.ts#L80-L187
- Adds integration tests for Find/Replace/Undo/Redo/Select All commands

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

- Find/Replace -> if the current editor then it should be used, if not then either suppressed or a widget should be used, i.e. the terminal
  - integration tests added to cover common cases
- Undo/Redo/Select All -> if the current editor then it should be used, otherwise delegate to DOM, rely on Monaco for implementation. Check input areas with and without editors in other views, git, SIW, terminal find, webview (gitlens welcome)
  - integration tests added to cover common cases
- Editor Actions like manipulation of selections
  - should be available in the command palette when the current editor is available and work properly
  - language features should be disable when they are not provided even for the current editor
  - test inline editors as debug console input and breakpoint expression editor
  - test embedded editors as peek references
- there should NOT be any exceptions in logs about missing service if the gitlens extension is installed
- integration tests should pass

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

akosyakov commented on packages/core/src/common/command.ts:
```diff
@@ -236,13 +236,17 @@ export class CommandRegistry implements CommandService {
     /**
      * Register the given handler for the given command identifier.
+     *
+     * If there is already a handler for the given command
```
it should go to CHANGELOG

kittaakos commented:
Executing `Undo` whilst the focus is an `input`; executes the command in the editor:

![screencast 2020-04-14 09-18-06](https://user-images.githubusercontent.com/1405703/79196879-2192fd80-7e31-11ea-9a60-934af61d5746.gif)

kittaakos commented:
Monaco editor commands incorrectly execute with the outer editor context, instead of the embedded one. Steps to reproduce:
 - Open an editor,
 - Open an embedded editor,
 - Set the focus in the inner editor,
 - Execute `Copy Line Up`, the command will copy a line based on the selection in the outer editor instead of the inner one.

![screencast 2020-04-14 09-25-08](https://user-images.githubusercontent.com/1405703/79197650-68352780-7e32-11ea-99dd-095448c278fc.gif)

akosyakov commented:
> Executing Undo whilst the focus is an input; executes the command in the editor:

You did not do any changes in the input, only focused it? I think it is the same in VS Code.

kittaakos commented:
> You did not do any changes in the input, only focused it? I think it is the same in VS Code.

It works in Code:
![screencast 2020-04-14 09-44-46](https://user-images.githubusercontent.com/1405703/79199111-c4994680-7e34-11ea-8765-df56bc729472.gif)

akosyakov commented:
> It works in Code:

@kittaakos in VS Code git input is a Monaco editor, try with the search in workspace input

kittaakos commented:
> @kittaakos in VS Code git input is a Monaco editor, try with the search in workspace input

You're right. The behavior isn't nice in VS Code editor, but Theia works as VS Code. I had to use the SIW `input` instead of the SCM ~`textArea`~ monaco-editor.

![screencast 2020-04-14 10-01-36](https://user-images.githubusercontent.com/1405703/79200837-5a35d580-7e37-11ea-9692-c1820912a6a1.gif)

akosyakov commented on packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:
```diff
@@ -193,18 +193,6 @@ export class PluginVscodeCommandsContribution implements CommandContribution {
         commands.registerCommand({ id: 'workbench.action.gotoLine' }, {
             execute: () => commands.executeCommand('editor.action.gotoLine')
         });
-        commands.registerCommand({ id: 'actions.find' }, {
-            execute: () => commands.executeCommand('actions.find')
-        });
-        commands.registerCommand({ id: 'undo' }, {
-            execute: () => commands.executeCommand('undo')
-        });
-        commands.registerCommand({ id: 'redo' }, {
-            execute: () => commands.executeCommand('redo')
-        });
-        commands.registerCommand({ id: 'editor.action.startFindReplaceAction' }, {
-            execute: () => commands.executeCommand('editor.action.startFindReplaceAction')
-        });
```
You should be able to call these commands via `CommandService`. I think we did it already by using undo/redo/select and find.

The issue with this code that it tries to override commands contributed by the monaco extension. In the good case it does nothing, just a warning in logs that such command already registered (what is happening on master i believe), in the worse it registers before monaco and causes infinity recursion.

akosyakov commented:
> Monaco editor commands incorrectly execute with the outer editor context, instead of the embedded one. Steps to reproduce:

You are right about getFocusedCodeEditor . It has to be checked as well before getActiveCodeEditor. Focused code editor includes inline and embedded editors, active code editors include only last focused standalone or diff editors.

I've adjuster handlers and added js-docs to exposed `getFocusedCodeEditor` to make it clear.

akosyakov commented on examples/api-tests/src/undo-redo-selectAll.spec.js:
```diff
@@ -0,0 +1,188 @@
+
+    afterEach(async () => {
+        toTearDown.dispose();
+        await scmContribution.closeView();
+        await navigatorContribution.closeView();
+        await editorManager.closeAll({ save: false });
```
During writing tests or running all of them together not everything always goes well and bogus termination of one test can break other tests. If it would be `beforeEach` should not be necessary.

akosyakov commented:
@kittaakos I've answered/addressed all your comments.

Dirty editor tests should be good now too. Code was assuming that we always load editor model to open from the disk or already processed all disk changes. But tests revealed that there are cases when we did not receive all events yet but already trying to open an editor. I've added a guard that if a model looks invalid then sync with the disk before throwing an invalid model exception.

I've also updated `How to test` section to reflect inline and embedded editors.

kittaakos commented:
@akosyakov, can you please rebase from the `master` HEAD? I want to verify #7523. Currently, it does not work.

akosyakov commented:
@kittaakos done, trying it as well as soon as the local build is through

kittaakos commented on packages/monaco/src/browser/monaco-editor-provider.ts:
```diff
@@ -105,6 +105,10 @@ export class MonacoEditorProvider {
 
     protected async getModel(uri: URI, toDispose: DisposableCollection): Promise<MonacoEditorModel> {
         const reference = await this.textModelService.createModelReference(uri);
+        // if document is invalid makes sure that all events from underlying resource are processed before throwing invalid model
+        if (!reference.object.valid) {
+            await reference.object.sync();
+        }
```
This change was not part of the PR originally; can you please give us some details why you've made this change? Is this related to the flaky tests or does something with the embedded code editor part? Thanks!

akosyakov replied:
It was added to fix an issue revealed by tests: https://github.com/eclipse-theia/theia/pull/7539#issuecomment-613444764

> Dirty editor tests should be good now too. Code was assuming that we always load editor model to open from the disk or already processed all disk changes. But tests revealed that there are cases when we did not receive all events yet but already trying to open an editor. I've added a guard that if a model looks invalid then sync with the disk before throwing an invalid model exception.

kittaakos replied:
👍 Thank you!

tavoda commented:
git pull (I'm on master)
yarn build
cd examples/browser
yarn start

Still same error. Using theia/plugins/vscodevim.vim-1.13.1.vsix with size 1952450. Do I have to do some 'clean' or using different vscode-vim plugin from other registry?

tavoda commented:
Found something at startup for CTRL+F:
root WARN A command editor.action.startFindReplaceAction is already registered.
root WARN Collided keybinding is ignored;  {"command":"passthrough","keybinding":"ctrl+f","context":"terminalActive"}  collided with  {"command":"terminal:find","keybinding":
"ctrlcmd+f","context":"terminalActive"}
root WARN Could not register keybinding:
  {"command":"passthrough","keybinding":"ctrl+f","context":"terminalActive"}
Error: "ctrl+f" is in collision with something else [scope:0]
root WARN Collided keybinding is ignored;  {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}  collided with  {"command":"terminal:clear","keybinding"
:"ctrlcmd+k","context":"terminalActive"}
root WARN Could not register keybinding:
  {"command":"passthrough","keybinding":"ctrl+k","context":"terminalActive"}
Error: "ctrl+k" is in collision with something else [scope:0]

akosyakov commented:
> A command editor.action.startFindReplaceAction is already registered.

@tavoda sounds like it was old version. This PR removes double registration:  https://github.com/eclipse-theia/theia/pull/7539#discussion_r407964812

tavoda commented:
Once more:
git pull
yarn run clean
yarn run build
cd examples/browser
yarn start -> ERROR?!?!?
yarn build
yarn start

EVERYTHING OK, CTRL + F is working :sunglasses: thank you
Now I can try to replace VSCODE with THEIA for everyday work.

tavoda commented:
Hmmm too soon, now basic movements like hjkl is not working, instead of movement it looks like in insert mode and &lt;ESC> doesn't help. However CTRL+F is working :cry:

akosyakov commented:
@tavoda thanks for checking, please open an issue 🙏
